File: code/richediting/src/main/java/org/fbme/ide/richediting/editor/RichResourceProjectionController.java
package org.fbme.ide.richediting.editor;

import com.intellij.openapi.actionSystem.AnAction;
import jetbrains.mps.project.Project;
import jetbrains.mps.smodel.ModelAccessHelper;
import org.fbme.ide.iec61499.adapter.fbnetwork.MPSNetworkInstanceReference;
import org.fbme.ide.iec61499.repository.PlatformRepository;
import org.fbme.ide.platform.PlatformRepositoryProvider;
import org.fbme.ide.platform.editor.ChooseProjectionAction;
import org.fbme.ide.platform.editor.EditorProjection;
import org.fbme.ide.platform.editor.EditorProjectionController;
import org.fbme.lib.iec61499.declarations.ResourceDeclaration;
import org.fbme.lib.iec61499.declarations.SystemDeclaration;
import org.fbme.lib.iec61499.instances.NetworkInstance;
import org.jdom.Element;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.mps.openapi.model.SNode;

import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class RichResourceProjectionController implements EditorProjectionController {

    private final SNode myNode;
    private final Project myProject;
    private final SystemDeclaration mySystem;


    public RichResourceProjectionController(SNode node, Project project) {
        myNode = node;
        myProject = project;
        mySystem = PlatformRepositoryProvider.getInstance(project).getAdapter(node, SystemDeclaration.class);
    }

    @NotNull
    @Override
    public String getId() {
        return "Resource";
    }

    @NotNull
    @Override
    public EditorProjection createProjection(@NotNull String name) {
        String[] names = name.split("\\.");
        final String deviceName = names[0];
        final String resourceName = names[1];
        final ResourceDeclaration resource = new ModelAccessHelper(myProject.getModelAccess()).runReadAction(() ->
                mySystem.getDevices().stream()
                        .filter(it -> Objects.equals(it.getName(), deviceName))
                        .findFirst()
                        .orElseThrow()
                        .getResources().stream()
                        .filter(it -> Objects.equals(it.getName(), resourceName))
                        .findFirst()
                        .orElseThrow()
        );
        NetworkInstance instance = NetworkInstance.createForResource(resource);
        return createProjection0(name, instance);
    }

    @Override
    public EditorProjection restoreProjection(@NotNull String name, @NotNull Element e) {
        PlatformRepository repository = PlatformRepositoryProvider.getInstance(myProject);

        String ref = e.getAttributeValue(RichApplicationProjection.PERSISTENCE_KEY);
        NetworkInstance instance = ref != null ? MPSNetworkInstanceReference.deserialize(ref).resolve(repository) : null;
        if (instance != null) {
            return createProjection0(name, instance);
        } else {
            return createProjection(name);
        }

    }

    private EditorProjection createProjection0(String name, NetworkInstance instance) {
        return new RichResourceProjection(myNode, this, name, new String[]{"org.fbme.ide.richediting.lang.editor.Rich Editing Hint.system_res"}, instance, myProject);
    }

    @Override
    public List<AnAction> getChooseProjectionActions() {
        return new ModelAccessHelper(myProject.getModelAccess()).runReadAction(() ->
                mySystem.getDevices().stream()
                        .flatMap(it -> it.getResources().stream())
                        .map(it -> new ChooseProjectionAction(this, it.getContainer().getName() + "." + it.getName()))
                        .collect(Collectors.toList())
        );
    }

    @Override
    public List<AnAction> getCreateProjectionActions() {
        return Collections.emptyList();
    }
}

File: code/richediting/src/main/kotlin/org/fbme/ide/richediting/editor/RichResourceProjectionController.kt
package org.fbme.ide.richediting.editor

import jetbrains.mps.project.Project
import jetbrains.mps.smodel.ModelAccessHelper
import org.fbme.ide.iec61499.fbnetwork.MPSNetworkInstanceReference
import org.fbme.ide.iec61499.repository.PlatformRepositoryProvider
import org.fbme.ide.platform.editor.ChooseProjectionAction
import org.fbme.ide.platform.editor.EditorProjection
import org.fbme.ide.platform.editor.EditorProjectionController
import org.fbme.ide.platform.editor.ProjectionChooser
import org.fbme.lib.iec61499.declarations.ResourceDeclaration
import org.fbme.lib.iec61499.declarations.SystemDeclaration
import org.fbme.lib.iec61499.instances.NetworkInstance
import org.jdom.Element
import org.jetbrains.mps.openapi.model.SNode

class RichResourceProjectionController(
        private val node: SNode,
        private val project: Project
) : EditorProjectionController {
    private val system: SystemDeclaration = PlatformRepositoryProvider
            .getInstance(project)
            .getAdapter(node, SystemDeclaration::class.java)
            ?: error("System declaration is null")

    override val id: String get() = "Resource"
    override val priority: Int get() = 1

    override val chooser: ProjectionChooser
        get() = ModelAccessHelper(project.modelAccess).runReadAction<ProjectionChooser> {
            ProjectionChooser.Composite(
                    id,
                    system.devices.flatMap { it.resources }
                            .map { ChooseProjectionAction(this, it.container.name + "." + it.name) },
                    emptyList()

            )
        }

    override fun createProjection(name: String): EditorProjection {
        val names = name.split('.').toTypedArray()
        val deviceName = names[0]
        val resourceName = names[1]
        val resource = ModelAccessHelper(project.modelAccess).runReadAction<ResourceDeclaration> {
            system.devices.first { it.name == deviceName }
                    .resources.first { it.name == resourceName }
        }
        val instance = NetworkInstance.createForResource(resource)
        return createProjection0(name, instance)
    }

    override fun restoreProjection(name: String, e: Element): EditorProjection {
        val repository = PlatformRepositoryProvider.getInstance(project)
        val ref = e.getAttributeValue(RichApplicationProjection.PERSISTENCE_KEY)
        val instance = if (ref != null) MPSNetworkInstanceReference.deserialize(ref).resolve(repository) else null
        return instance?.let { createProjection0(name, it) } ?: createProjection(name)
    }

    private fun createProjection0(name: String, instance: NetworkInstance): EditorProjection {
        return RichResourceProjection(
                node,
                this,
                name,
                arrayOf("org.fbme.ide.richediting.lang.editor.Rich Editing Hint.system_res"),
                instance,
                project
        )
    }
}
